frontend/controller/actions/identity.js

```diff
@@ -701,5 +701,72 @@ export default (sbp('sbp/selectors/register', {
   ...encryptedAction('gi.actions/identity/joinDirectMessage', L('Failed to join a direct message.')),
   ...encryptedAction('gi.actions/identity/joinGroup', L('Failed to join a group.')),
   ...encryptedAction('gi.actions/identity/leaveGroup', L('Failed to leave a group.')),
-  ...encryptedAction('gi.actions/identity/setDirectMessageVisibility', L('Failed to set direct message visibility.'))
+  ...encryptedAction('gi.actions/identity/setDirectMessageVisibility', L('Failed to set direct message visibility.')),
+  'gi.actions/identity/uploadFiles': async ({ attachments, billableContractID }: {
+    attachments: Object, billableContractID: string
+  }) => {
+    const rootGetters = sbp('state/vuex/getters')
+
+    try {
+      const attachmentsData = await Promise.all(attachments.map(async (attachment) => {
+        const { mimeType, url } = attachment
+        // url here is an instance of URL.createObjectURL(), which needs to be converted to a 'Blob'
+        const attachmentBlob = await objectURLtoBlob(url)
+        const response = await sbp('chelonia/fileUpload', attachmentBlob, {
+          type: mimeType, cipher: 'aes256gcm'
+        }, { billableContractID })
+        const { delete: token, download: downloadData } = response
+        return {
+          downloadData: { ...omit(attachment, ['url']), downloadData },
+          deleteData: { token }
+        }
+      }))
+
+      const tokensByManifestCid = attachmentsData.map(({ downloadData, deleteData }) => ({
+        manifestCid: downloadData.downloadData.manifestCid,
+        token: deleteData.token
+      }))
+
+      await sbp('gi.actions/identity/saveFileDeleteToken', {
+        contractID: rootGetters.ourIdentityContractId,
+        data: { tokensByManifestCid }
+      })
+
+      return attachmentsData.map(({ downloadData }) => downloadData)
+    } catch (err) {
+      console.error('Error during uploading files', err)
+    }
+  },
+  'gi.actions/identity/removeFiles': async ({ manifestCids, identityContractID }: {
+    manifestCids: string[], identityContractID: string, billableContractID?: string
+  }) => {
+    const rootGetters = sbp('state/vuex/getters')
+    const me = sbp('state/vuex/state').loggedIn.identityContractID
+
+    if (me !== identityContractID) {
+      return
+    }
+
+    try {
+      const tokensMap = {}
+      for (const manifestCid of manifestCids) {
+        const token = rootGetters.currentIdentityState.fileDeleteTokens[manifestCid]
+        if (!token) {
+          console.error(`Missing delete token for file with manifestCid ${manifestCid}`)
```

Comment:
Then, what about using `console.warn`? And updated the logic according to your feedback. Please check it out.
```
'gi.actions/identity/removeFiles': async ({ manifestCids, identityContractID }: {
  manifestCids: string[], identityContractID: string, billableContractID?: string
}) => {
  const rootGetters = sbp('state/vuex/getters')
  const me = sbp('state/vuex/state').loggedIn.identityContractID

  if (me !== identityContractID) {
    return
  }

  try {
    const tokensMap = {}
    for (const manifestCid of manifestCids) {
      const token = rootGetters.currentIdentityState.fileDeleteTokens[manifestCid]
      if (!token) {
        console.warn(`Missing delete token for file with manifestCid ${manifestCid}`)
      } else {
        tokensMap[manifestCid] = { token }
      }
    }
    const removableManifestCids = Object.keys(tokensMap)

    await sbp('chelonia/fileDelete', removableManifestCids, tokensMap)
    await sbp('gi.actions/identity/removeFileDeleteToken', {
      contractID: identityContractID,
      data: { manifestCids: removableManifestCids }
    })
  } catch (err) {
    console.warn(err?.message || err)
  }
}
```

Reply:
I think using `console.warn` is fine here, but also see @taoeffect's comment on error handling. _Ideally_ failing to delete a file _would_ be treated as an error, but only when the user is the one initiating the action. One way to do this could be to do the deletion as part of the `action` instead of as part of the contract (OTOH, doing it in the action makes it more fault tolerant).